decred/tests/integration/util/test_ws.py

```diff
@@ -0,0 +1,98 @@
+"""
+Copyright (c) 2020, the Decred developers
+See LICENSE for details
+"""
+
+from queue import Queue
+import threading
+import time
+
+import pytest
+import websocket
+from websocket_server import WebsocketServer
+
+from decred.util import ws
+
+
+HOST = "localhost"
+PORT = 53791
+URL = f"http://{HOST}:{PORT}"
+QUIT = "QUIT"
+HELLO = "HELLO"
+YO = "YO"
+CLOSE_HANDSHAKE = bytearray([0x88, 0])
+
+
+def test_websocket():
+    """
+    Start a ws.Client and check basic functionality.
+    """
+    serverQ = Queue(1)
+    clientQ = Queue(1)
+    errorQ = Queue(1)
+
+    def receive(client, server, message):
+        if message == HELLO:
+            server.send_message(client, YO)
+        elif message == QUIT:
+            client["handler"].request.send(CLOSE_HANDSHAKE)
+            server.shutdown()
+        serverQ.put(message)
+
+    server = WebsocketServer(PORT, HOST)
+    server.set_fn_message_received(receive)
+    serverThread = threading.Thread(None, server.run_forever)
+    serverThread.start()
+    time.sleep(1)
```

Comment:
I hate doing this, but there is no obvious way to tie into any type of "server ready" signal from either `WebSocketServer` or the underlying `TCPServer`.